### DISCUSSION TITLE
How to mock Promise methods

### DISCUSSION THREAD

**Discussion by sleonia** [vitest-dev/vitest#3778] [Q&A] [answered]
Hi. I have a leetcode-like task and I want to limit students using Promise methods. I want to prevent using `Promise.all`, `Promise.race`, `Promise.any` and `Promise.allSettled`. 

I have this code

```javascript
import { describe, it, expect, afterEach, beforeEach, vi } from "vitest"

describe("Example", () => {
  const allSpy = vi.spyOn(Promise, "all")
  const allSettledSpy = vi.spyOn(Promise, "allSettled")
  const anySpy = vi.spyOn(Promise, "any")
  const raceSpy = vi.spyOn(Promise, "race")


  beforeEach(() => {
    vi.resetAllMocks();
  });

  afterEach(() => {
    expect(allSpy).not.toBeCalled()
    expect(allSettledSpy).not.toBeCalled()
    expect(anySpy).not.toBeCalled()
    expect(raceSpy).not.toBeCalled()
  })

  it("runs my program", () => { /* 👻 */ })
}
```

When I run it I see that some Promise methods been called inside runner and my expects are failing.
https://github.com/vitest-dev/vitest/blob/a90d64fee3d07037ffccac54c032603cca00d10a/packages/runner/src/run.ts#L54
I was thinking that my mocks affects only testing program, not vite runners.

How to mock Promise methods correctly?

My vitest version `"^0.32.2"`

**Comment by sheremet-va** [comment, marked as answer]
I don't think you can without a rewrite inside Vitest. You can open a separate issue for this.
